refactor: Convert TableElement.jsx component from class to functional with hooks

corbinrobb opened the pull request:
### SUMMARY
Update the TableElement class component to be a functional component using hooks.

Modified the tests slightly where needed to check the changes to DOM from updating the state instead of checking the component state itself. Was necessary because functional components using hooks don't have an accessible state like class components.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
The component should look and function the same

### TESTING INSTRUCTIONS
To run the tests for TableElement navigate to superset-frontend and run `npm run test -- spec/javascripts/sqllab/TableElement_spec.jsx`

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
Change TableElement from a class component to a functional component

Commit:
Replace class state checks in TableElement_spec.jsx with checks testing elements they change

lyndsiWilliams commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -253,33 +243,31 @@ class TableElement extends React.PureComponent {
+        {renderWell()}
         <div>
           {cols &&
             cols.map(col => <ColumnElement column={col} key={col.name} />)}
```
This can be simplified to `{cols?.map(col => <ColumnElement column={col} key={col.name} />)}`

corbinrobb replied:
Sending up a commit!

Commit:
Refactor small bit of logic to use optional chaining

lyndsiWilliams commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -126,11 +119,10 @@ class TableElement extends React.PureComponent {
+  const renderControls = () => {
     let keyLink;
-    const { table } = this.props;
     if (table.indexes && table.indexes.length > 0) {
```
This can also be simplified to just `(table?.indexes)`

corbinrobb replied:
Pushed up changes!

Commit:
Add optional chaining to some logic

Commit:
Fix IconTooltip and add IconTooltip to the collapse button

Commit:
Fix custom icon using IconToolTip so it better matches the original

corbinrobb commented:
***
Figured it would make sense to grab a ticket that is also making changes in this component.

Please let me know if this should be in another PR or if there's a better way to handle it. Thanks!
***

## [sql lab]'copy' and 'collapse/expand' icons do not have tooltip in the table review panel

### SUMMARY
Fixed an issue where the tooltip for the copy/paste was not showing up and added a tooltip to the collapse/expand icon on the table element.

Updated tests where necessary.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

![image](https://user-images.githubusercontent.com/31329271/119869249-3d464f80-bedd-11eb-9a91-e89c3097b2f6.png)### TESTING INSTRUCTIONS
To run the tests for TableElement navigate to superset-frontend and run `npm run test -- spec/javascripts/sqllab/TableElement_spec.jsx`

![table-element](https://user-images.githubusercontent.com/31329271/119870893-12f59180-bedf-11eb-8ee3-301ff8cfffc8.gif)



### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

AAfghahi commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -32,6 +32,7 @@ import ColumnElement from './ColumnElement';
 import ShowSQL from './ShowSQL';
 import ModalTrigger from '../../components/ModalTrigger';
 import Loading from '../../components/Loading';
+import { RightOutlined } from '@ant-design/icons';
```
Hey, we actually use this for our Icons:
https://github.com/apache/superset/blob/master/superset-frontend/src/components/Icon/index.tsx

THis is a place where we use it which might help:
https://github.com/apache/superset/blob/master/superset-frontend/src/views/CRUD/data/query/QueryList.tsx

corbinrobb replied:
I see, thank you I will update that as soon as I can!

AAfghahi commented:
Heya, for the testing instructions, I think that adding how to visually test for it might be best. So something short and along the lines of ---> navigate to "x" then whatever would be good. So that someone who pulls your file, or if they create an ephemeral env could go and visually inspect it.

eschutho commented:
@yousoph how do these tooltips look?

yousoph commented:
Nice! Looking good to me, thanks @corbinrobb!

Commit:
Update collapse/expand icon to use Icons component instead of importing from antdesign directly

corbinrobb commented:
It was my pleasure! @yousoph I'm looking forward to helping more!

hughhhh commented:
@corbinrobb can you fix these issues to allow CI to pass
```
/home/runner/work/superset/superset/superset-frontend/src/SqlLab/components/TableElement.jsx
   35:1   error  `src/components/Icons` import should occur before import of `../../components/CopyToClipboard`         import/order
   70:9   error  'popSelectStar' is assigned a value but never used                                                     no-unused-vars
  192:30  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
  264:36  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
  280:20  error  Curly braces are unnecessary here                                                                      react/jsx-curly-brace-presence
  292:7   error  Value must be omitted for boolean attributes                                                           react/jsx-boolean-value
```

AAfghahi commented:
> @corbinrobb can you fix these issues to allow CI to pass
> 
> ```
> /home/runner/work/superset/superset/superset-frontend/src/SqlLab/components/TableElement.jsx
>    35:1   error  `src/components/Icons` import should occur before import of `../../components/CopyToClipboard`         import/order
>    70:9   error  'popSelectStar' is assigned a value but never used                                                     no-unused-vars
>   192:30  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
>   264:36  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
>   280:20  error  Curly braces are unnecessary here                                                                      react/jsx-curly-brace-presence
>   292:7   error  Value must be omitted for boolean attributes                                                           react/jsx-boolean-value
> ```

So I actually think popSelectStar is depreciated. Might be worth setting up a test env to check to make sure.

Commit:
Fix eslint errors

corbinrobb commented:
@hughhhh I got the linter errors figured out and I pushed up the latest commit. I agree with @AAfghahi about popSelectStar. I removed it from the component since it was throwing the unused variable linter error. It appears it wasn't being used before I started the PR either but not throwing the linter because it was a method on a class. Thanks lemme know if I should/need to do anything else with this!

hughhhh commented:
/testenv up

eschutho commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -126,12 +108,11 @@ class TableElement extends React.PureComponent {
     let keyLink;
-    const { table } = this.props;
-    if (table.indexes && table.indexes.length > 0) {
+    if (table?.indexes.length > 0) {
```
if (table?.indexes?.length)

0 will be falsy, and it won't be negative. Also looks like indexes could be null or undefined.

eschutho replied:
a lot of people still prefer foo.length > 0, so it's not wrong, just a personal preference. :)

eschutho commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -156,26 +137,28 @@ class TableElement extends React.PureComponent {
         <IconTooltip
           className={
-            `fa fa-sort-${!this.state.sortColumns ? 'alpha' : 'numeric'}-asc ` +
+            `fa fa-sort-${!sortColumns ? 'alpha' : 'numeric'}-asc ` +
```
small nit, and this is from earlier code, but for readability: 
``fa fa-sort-${sortColumns ?  'numeric' : 'alpha' }-asc ` +`

eschutho commented on superset-frontend/src/SqlLab/components/TableElement.jsx:
```diff
@@ -253,33 +232,54 @@ class TableElement extends React.PureComponent {
+        alignItems: 'center',
+      }}
+      aria-label="Collapse"
+      tooltip={t(`${!isActive ? 'Expand' : 'Collapse'} table preview`)}
```
can you do the same thing here?
tooltip={t(`${isActive ?  'Collapse' : 'Expand'} table preview`)}

eschutho commented:
This is great @corbinrobb, thanks so much! I just left a few nits, and some of them are from earlier code that I thought I'd see if we could sneak in. :)

corbinrobb commented:
No problem at all! @eschutho I like all the changes! Much more concise and readable. I will be pushing them up in just a moment.

Commit:
Clean up some code for readability

betodealmeida reviewed (approved):
<img src="https://media.giphy.com/media/l2RsBhyCSu5Y5XU3Zf/giphy.gif" />
